Twig/NavigationExtension.php

```diff
@@ -76,19 +179,50 @@ public function getNavigationActiveAtLevel($level)
      * Finds a node indexed in the tree. See \Webfactory\Bundle\NavigationBundle\Tree\Tree::find.
      *
      * @param array $provisions parameters used to look up the node
-     *
-     * @return \Webfactory\Bundle\NavigationBundle\Tree\Node|null
      */
-    public function findNode(array $provisions)
+    public function findNode(array $provisions): ?Node
     {
         return $this->getTree()->find($provisions);
     }
 
-    /**
-     * @return Tree
-     */
-    private function getTree()
+    public function getAdditionalNavigationItemClasses(Node $node, array $loop, int $level): string
+    {
+        $baseClasses = [
+            $node->isActiveNode() ? 'a' : 'na',
```

Comment:
Hm, das packt irgendwelche CSS-Klassennamen in die Twig-Extension, wo sie bestimmt nicht gut auszutauschen sind... :-(. Muss das sein?

Wo lebte das vorher? Was das der Grund, warum das "power set" als Funktion angeboten wurde, damit die Ableitung dieser CSS-Klassen-Kombinationen im Theme passieren konnte?

Reply:
Vorher im Twig Template fürs Theme: https://github.com/webfactory/WebfactoryNavigationBundle/blob/f55fa0cd7f46ee8628f30091b01559060660a6b7/Resources/views/Navigation/baseNavigationTheme.html.twig#L29-L45

Ich _vermute_, dass Team Grün entweder alle diese Klassen fürs Styling möchte - und dann der Einheitlichkeit halber in genau dieser Schreibweise, oder keine - dann kann es den Block

```
{% block navigation_item_class -%}
    ni ni-{{ level }} {{ additional_navigation_item_classes(node, loop, level) }}
{%- endblock -%}
```

entsprechend überschreiben. Oder wollt ihr wie vorher (siehe Link oben) möglich, mit der power_set-Funktion neu benannte Klassen bauen können, @polarbirke @megatronCGN?

Reply:
Tatsächlich sind wir von den kryptischen Kürzeln seit ~2 Jahren weg und tendieren eher dazu, im projekt-spezifischen Theme verbose Schreibweise inkl. Projekt-Namespace/Präfix zu definieren.

Beispiel: https://github.com/webfactory/jugendfuereuropa/blob/98014_Redesign/src/Webforum/JFE/Bundle/WebsiteBundle/Resources/views/Navigation/main_navigation_theme.html.twig#L68 ff.

Reply:
Die Idee war mal in grauer Vorzeit, Klassen und -kombinationen "für alles" zu haben, so dass man Styling auf ein Element anwenden konnte, wenn es "das aktive Element auf 2. Ebene ist, das nicht das letzte unter seinen Geschwistern darstellt" oder so ähnlich. Es wurde tatsächlich immer nur ein Bruchteil der Potenzmenge (= alle möglichen Kombinationen der Klassen) gebraucht.

Worum es mir geht: Entweder, wir behalten das Feature für BC bei, oder wir machen an der Stelle einen break, lassen die Klassen raus und man muss im Projekt selbst definieren, was man will... das geht sicher auch ohne "power set". 

Verstehe ich @MalteWunsch richtig, dass Du das `power_set` _nicht_ mehr als Twig-Funktion anbieten möchtest, dann aber – um noch die alten Klassen für BC liefern zu können – den Krempel direkt in der Twig-Extension hast?

Wenn dann aber jemand die Klassen überschrieben hat und dafür das `power_set` braucht, dann haben wir nichts gewonnen... die BC ist dann trotzdem gebrochen, und die CSS-Klassen im Extension-Code sind auch wertlos. 🤷🏻‍♂️

Reply:
Potzblitz: [wir haben tatsächlich Klassen mit Hilfe von `power_set` überschrieben](https://github.com/search?q=org%3Awebfactory+power_set&type=code), und das nicht nur einmal. Das hatte ich gar nicht in Betracht gezogen. Alles klar, dann bleibt die `power_set` als TwigFunction erhalten!

Dann schwenke ich auch auf die alte Implementierung des `{% block navigation_item_class -%}`-Blocks zurück. Ich fand die zwar schwerer lesbar, bin da aber sicherlich gebiased. Außerdem gefiel mir auch nicht, dafür eine TwigFunction `additional_navigation_item_classes` public anbieten zu müssen.